src/parquet/arrow/reader.cc

```diff
@@ -265,61 +266,125 @@ void FlatColumnReader::Impl::ReadNonNullableBatch<::arrow::BooleanType, BooleanT
+  valid_bits_idx_ += *levels_read;
+
+  return Status::OK();
 }
 
+#define NULLABLE_BATCH_FAST_PATH(ArrowType, ParquetType, CType)                      \
+  template <>                                                                        \
+  Status FlatColumnReader::Impl::ReadNullableFlatBatch<ArrowType, ParquetType>(      \
+      TypedColumnReader<ParquetType> * reader, int16_t * def_levels, CType * values, \
+      int64_t values_to_read, int64_t * levels_read) {                               \
+    auto data_ptr = reinterpret_cast<CType*>(data_buffer_ptr_);                      \
+    int null_count;                                                                  \
+    PARQUET_CATCH_NOT_OK(*levels_read = reader->ReadBatchSpaced(values_to_read,      \
+                             def_levels, nullptr, data_ptr + valid_bits_idx_,        \
+                             &null_count, valid_bits_ptr_, valid_bits_idx_));        \
+                                                                                     \
+    valid_bits_idx_ += *levels_read;                                                 \
+    null_count_ += null_count;                                                       \
+                                                                                     \
+    return Status::OK();                                                             \
+  }
+
+NULLABLE_BATCH_FAST_PATH(::arrow::Int32Type, Int32Type, int32_t)
+NULLABLE_BATCH_FAST_PATH(::arrow::Int64Type, Int64Type, int64_t)
+NULLABLE_BATCH_FAST_PATH(::arrow::FloatType, FloatType, float)
+NULLABLE_BATCH_FAST_PATH(::arrow::DoubleType, DoubleType, double)
+
 template <>
-void FlatColumnReader::Impl::ReadNullableFlatBatch<::arrow::TimestampType, Int96Type>(
-    const int16_t* def_levels, Int96* values, int64_t values_read, int64_t levels_read) {
+Status FlatColumnReader::Impl::ReadNullableFlatBatch<::arrow::TimestampType, Int96Type>(
+    TypedColumnReader<Int96Type>* reader, int16_t* def_levels, Int96* values,
+    int64_t values_to_read, int64_t* levels_read) {
+  int null_count;
+  PARQUET_CATCH_NOT_OK(*levels_read =
+                           reader->ReadBatchSpaced(values_to_read, def_levels, nullptr,
+                               values, &null_count, valid_bits_ptr_, valid_bits_idx_));
+
   auto data_ptr = reinterpret_cast<int64_t*>(data_buffer_ptr_);
-  int values_idx = 0;
-  for (int64_t i = 0; i < levels_read; i++) {
-    if (def_levels[i] < descr_->max_definition_level()) {
-      null_count_++;
-    } else {
-      ::arrow::BitUtil::SetBit(valid_bits_ptr_, valid_bits_idx_);
-      data_ptr[valid_bits_idx_] = impala_timestamp_to_nanoseconds(values[values_idx++]);
+  int byte_offset = valid_bits_idx_ / 8;
+  int bit_offset = valid_bits_idx_ % 8;
+  uint8_t bitset = valid_bits_ptr_[byte_offset];
+  for (int64_t i = 0; i < *levels_read; i++) {
+    if (bitset & (1 << bit_offset)) {
+      data_ptr[valid_bits_idx_ + i] = impala_timestamp_to_nanoseconds(values[i]);
```

Comment:
You could define an inline "transfer function" of sorts, e.g. (just thinking out loud here)

```c++
template <>
struct ValueTransfer<Int96, int64_t> {
  static inline void Transfer(const Int96& value, int64_t* out) {
    *out = impala_timestamp_to_nanoseconds(value);
  }
};
```